test/e2e/gameserver_test.go

```diff
@@ -213,6 +213,113 @@ func TestGameServerUnhealthyAfterDeletingPod(t *testing.T) {
+	err = wait.PollImmediate(3*time.Second, 3*time.Minute, func() (bool, error) {
+		pod, err := framework.KubeClient.CoreV1().Pods(defaultNs).Get(newGs.ObjectMeta.Name, metav1.GetOptions{})
+		if err != nil {
+			logger.WithError(err).Warn("could not get pod for gameserver")
+			return true, err
+		}
+
+		for _, c := range pod.Status.ContainerStatuses {
+			if c.Name == newGs.Spec.Container && c.RestartCount >= 1 {
+				logger.Info("successfully crashed. Moving on!")
+				return true, nil
+			}
+		}
+
+		// create a connection each time, as weird stuff happens if the receiver isn't up and running.
+		conn, err := net.Dial("udp", address)
+		assert.NoError(t, err)
+		defer conn.Close() // nolint: errcheck
+		// doing this last, so that there is a short delay between the message being sent, and the check.
+		// also gives the process a chance to come back up
+		_, err = conn.Write([]byte("CRASH"))
+		logger.WithError(err).Info("no crash, trying again")
+		return false, nil
+	})
+	assert.NoError(t, err)
+
+	// check that the GameServer is not in an unhealthy state. If it does happen, it should happen pretty quick
+	newGs, err = framework.WaitForGameServerState(newGs, agonesv1.GameServerStateUnhealthy, 5*time.Second)
+	// should be an error, as the state should not occur
+	if !assert.Error(t, err) {
+		assert.FailNow(t, "GameServer should not be Unhealthy")
+	}
+	assert.Contains(t, err.Error(), "waiting for GameServer")
+
+	// ping READY until it doesn't fail anymore - since it may take a while
+	// for this to come back up -- or we could get a delayed CRASH, so we have to
+	// wait for the process to restart again to fire the SDK.Ready()
+	logger.Info("marking GameServer as ready")
+	err = wait.PollImmediate(3*time.Second, 3*time.Minute, func() (bool, error) {
+		newGs, err = gsClient.Get(newGs.ObjectMeta.Name, metav1.GetOptions{})
+		if err != nil {
+			logger.WithError(err).Warn("could not get gameserver")
+			return true, err
+		}
+
+		if newGs.Status.State == agonesv1.GameServerStateReady {
+			logger.Info("ready! Moving On!")
+			return true, nil
+		}
+
+		// create a connection each time, as weird stuff happens if the receiver isn't up and running.
+		conn, err := net.Dial("udp", address)
+		assert.NoError(t, err)
+		defer conn.Close() // nolint: errcheck
+		// doing this last, so that there is a short delay between the message being sent, and the check.
+		_, err = conn.Write([]byte("READY"))
+		logger.WithError(err).WithField("gs", newGs.ObjectMeta.Name).WithField("state", newGs.Status.State).Info("no ready, trying again")
+		return false, nil
+	})
+	if !assert.NoError(t, err) {
+		assert.FailNow(t, "Could not make GameServer Ready")
+	}
+	// now crash, should be unhealthy, since it's after being Ready
+	logger.Info("crashing again, should be unhealthy")
+	// retry on crash, as with the restarts, sometimes Go takes a moment to send this through.
+	err = wait.Poll(time.Second, 3*time.Minute, func() (bool, error) {
```

Comment:
All three of these udp sending loops are very similar in structure (write a packet, check for state, try again). Have you considered moving them into a helper function?

Reply:
Nice. Done!